derive/src/trait_bounds.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloc::vec::Vec;
+use proc_macro2::Ident;
+use syn::{
+    parse_quote,
+    spanned::Spanned,
+    visit::Visit,
+    Generics,
+    Result,
+    Type,
+};
+
+/// Visits the ast and checks if one of the given idents is found.
+struct ContainIdents<'a> {
+    result: bool,
+    idents: &'a [Ident],
+}
+
+impl<'a, 'ast> Visit<'ast> for ContainIdents<'a> {
+    fn visit_ident(&mut self, i: &'ast Ident) {
+        if self.idents.iter().any(|id| id == i) {
+            self.result = true;
+        }
+    }
+}
+
+/// Checks if the given type contains one of the given idents.
+fn type_contains_idents(ty: &Type, idents: &[Ident]) -> bool {
+    let mut visitor = ContainIdents {
+        result: false,
+        idents,
+    };
+    visitor.visit_type(ty);
+    visitor.result
+}
```

Comment:
I would prefer the `add` function below to be the top of the file as the entry point, helpers below. Also this visitor might be better to be inline inside the only function where it is used, or at least just below it for easy reference